core/model/src/main/java/org/eclipse/rdf4j/model/impl/DynamicModel.java

```diff
@@ -141,6 +142,35 @@ public boolean remove(Resource subj, IRI pred, Value obj, Resource... contexts)
 
 	@Override
 	public Model filter(Resource subj, IRI pred, Value obj, Resource... contexts) {
+		if (model == null && statements.size() == 0) {
+			return new EmptyModel(this);
+		}
+
+		if (model == null && subj != null && pred != null && obj != null && contexts.length == 1) {
+			Statement needle = SimpleValueFactory.getInstance().createStatement(subj, pred, obj, contexts[0]);
+			if (statements.size() == 1) {
+				Statement s = (Statement) statements.toArray()[0];
+
+				DynamicModel ret = new DynamicModel(modelFactory);
+
+				if (s.equals(needle)) {
+					ret.add(s);
+				}
+
+				return ret;
+			}
+
+			DynamicModel ret = new DynamicModel(modelFactory);
+			if (statements.contains(needle)) {
+				ret.add(needle);
+			}
+			return ret;
+		}
+
+		if (subj == null && pred == null && obj == null && contexts.length == 0) {
+			return this;
+		}
+
 		upgrade();
 		return model.filter(subj, pred, obj, contexts);
 	}
```

Comment:
In order to get the most out of sails that use the notifying behaviour we need to implement simple support for filter. I read that the docs specify that filter(...) should return a view of the current model, rather than a new model or an unmodifiable model. Not sure how to implement this with the performance I want. Have you worked with the FilteredModel before @jeenbroekstra ?

Reply:
I'm not sure what problem you're trying to solve here, to be honest. Just filter on the underlying model when filter is used. It will use a FilteredModel internally.

Reply:
I haven't thought this through fully, but if your idea is to somehow postpone upgrading when filtering by just using the original DynamicModel in corner cases, you could just do this by returning a new FilteredModel that internally creates an iterator, and put your filtering / upgrade logic in that iterator. As long as you make sure that what you return is a filtered view on the original Model, so that any future modifications of that result object get reflected in the original model, you're good.